src/routes/run.rs

```diff
@@ -32,14 +40,199 @@ pub struct SubmitTravel {
+pub struct CorrelatePlease {
+    pub run_id: Uuid,
+    pub get_result: bool,
+    pub get_stats: bool,
+}
+
+/// Response to explicit correlate request
+#[derive(Serialize, Deserialize, ToSchema)]
+pub struct CorrelateResponse {
+    pub new_report_points: i64,
+    pub updated_report_points: i64,
+    pub stops_file: Option<LocationsJson>,
+}
+
+/// This endpoint would correlate runs for given user id. For regular user only own runs
+/// can be correlated, for admin - any run for any user
+#[utoipa::path(
+    post,
+    path = "/correlate/run",
+    responses(
+        (status = 200, description = "Correlation Successful", body = CorrelateResponse),
+        (status = 401, description = "Unauthorized"),
+        (status = 403, description = "Forbidden"),
+        (status = 500, description = "Interal Error"),
+        (status = 501, description = "Not Implemented"),
+    ),
+)]
+pub async fn correlate_run(
+    pool: web::Data<DbPool>,
+    user: Identity,
+    _req: HttpRequest,
+    corr_request: web::Json<CorrelatePlease>,
+) -> Result<web::Json<CorrelateResponse>, ServerError> {
+    if corr_request.get_stats {
+        return Err(ServerError::NotImplemented);
+    }
+
+    // get connection from the pool
+    let mut database_connection = match pool.get() {
+        Ok(conn) => conn,
+        Err(e) => {
+            error!("cannot get connection from connection pool {:?}", e);
+            return Err(ServerError::InternalError);
+        }
+    };
+
+    // Parse the user id
+    let uuid: Uuid = match user.id() {
+        Ok(id) => match Uuid::parse_str(&id) {
+            Ok(uid) => uid,
+            Err(e) => {
+                error!("While parsing user UUID: {e}");
+                return Err(ServerError::Unauthorized);
+            }
+        },
+        Err(e) => {
+            error!("While trying to read user id from request: {e}");
+            return Err(ServerError::Unauthorized);
+        }
+    };
+
+    // fetch the user
+    use tlms::schema::users::dsl::users;
+    use tlms::schema::users::id as user_id;
+
+    // Get the user and privileges
+    let req_user: User = match users
+        .filter(user_id.eq(uuid))
+        .first(&mut database_connection)
+    {
+        Ok(user) => user,
+        Err(e) => {
+            error!("While trying to query user info for {uuid}: {e}");
+            return Err(ServerError::Unauthorized);
+        }
+    };
+
+    use tlms::schema::trekkie_runs::dsl::trekkie_runs;
+    use tlms::schema::trekkie_runs::id as run_id;
+    let run: TrekkieRun = match trekkie_runs
+        .filter(run_id.eq(corr_request.run_id))
+        .first(&mut database_connection)
+    {
+        Ok(r) => r,
+        Err(e) => {
+            error!("While trying to query for run {}: {e}", corr_request.run_id);
+            return Err(ServerError::InternalError);
+        }
+    };
+
+    if run.owner != req_user.id && req_user.role != Role::Administrator as i32 {
```

Comment:
Maybe smart to give stats for the owner and writing into storage exclusively for admins

Reply:
I would shift this problem to future us, when we move stops.json into the db